### PR TITLE
[Strings] Represent string values as WTF-16 internally

### PR DESCRIPTION
**Pull request by tlively**
WTF-16, i.e. arbitrary sequences of 16-bit values, is the encoding of Java and
JavaScript strings, and using the same encoding makes the interpretation of
string operations trivial, even when accounting for non-ascii characters.
Specifically, use little-endian WTF-16.

Re-encode string constants from WTF-8 to WTF-16 in the parsers, then back to
WTF-8 in the writers. Update the constructor for string `Literal`s to interpret
the string as WTF-16 and store a sequence of WTF-16 code units, i.e. 16-bit
integers. Update `Builder::makeConstantExpression` accordingly to convert from
the new `Literal` string representation back to a WTF-16 string.

Update the interpreter to remove the logic for detecting non-ascii characters
and bailing out. The naive implementations of all the string operations are
correct now that our string encoding matches the JS string encoding.

### CODE REVIEW

**Comment by tlively** [comment]
* **#6418** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/6418?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlively and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: CHANGELOG.md
Comment (tlively) [reply CHANGELOG.md:24]: It shouldn't be, no. Before this change, most invalid WTF-8 would end up littered with replacement characters in the output anyway, and certainly no one wants that.

File: src/wasm-builder.h
Comment (tlively) [reply src/wasm-builder.h:1276]: Yes, so we're not doing an encoding conversion here, but rather a data structure conversion. The `Literal` stores a sequence of `Literal` integers and we need to convert that to a `Name`.

File: test/lit/passes/string-lowering.wast
Comment (tlively) [reply test/lit/passes/string-lowering.wast:30]: Correct. These test cases are all invalid WTF-8 and we were previously testing that we printed warnings and inserted replacement characters correctly, but now we just reject them in the parsers.

File: test/lit/strings.wast
Comment (tlively) [reply test/lit/strings.wast:45]: Oh right, for the JSON escaping. I'll add that back in.

**Review by kripken — APPROVED** [review]
lgtm, modulo UTF8/16 nuances that I am not an expert on... but the fuzzer should now be able to test such things by comparing to V8. Might be worth fuzzing this for a while before landing.

File: src/wasm-interpreter.h
Comment (kripken) [inline src/wasm-interpreter.h:2139]: The previous line ensures that `endVal <= refValues.size()`, so this check is always false.

**Review by kripken — APPROVED** [review]

**Pull request merged**
